modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/RebalanceStatisticsUtils.java

```diff
@@ -79,536 +67,267 @@ private RebalanceStatisticsUtils() {
+            .a(grpInfo(cacheGrpCtx)).a(", ").a(time(stat.start(), stat.end())).a(", restarted=")
+            .a(stat.attempt() - 1).a("] ");
 
-        StringJoiner joiner = new StringJoiner(" ");
+        Map<ClusterNode, CacheGroupSupplierRebalanceStatistics> supStats = stat.supplierStatistics();
+        if (supStats.isEmpty())
+            return sb.toString();
 
-        if (finish)
-            writeTotalRebalanceStatistics(rebFutrs, nodeAliases, joiner);
+        sb.a(SUP_STAT_HEAD);
 
-        writeCacheGroupsRebalanceStatistics(rebFutrs, nodeAliases, finish, joiner);
-        writeAliasesRebalanceStatistics("p - partitions, e - entries, b - bytes, d - duration", nodeAliases, joiner);
-        writePartitionsDistributionRebalanceStatistics(rebFutrs, nodeAliases, nodeCnt, joiner);
+        int nodeId = 0;
+        for (CacheGroupSupplierRebalanceStatistics supStat : supStats.values()) {
+            long fp = 0, hp = 0;
 
-        return joiner.toString();
-    }
+            for (Entry<Integer, Boolean> pe : supStat.partitions().entrySet()) {
+                if (pe.getValue())
+                    fp++;
+                else
+                    hp++;
+            }
 
-    /**
-     * Write total statistics for rebalance.
-     *
-     * @param rebFutrs Participating in successful and not rebalances, require not null.
-     * @param nodeAliases For print nodeId=1 instead long string, require not null.
-     * @param joiner For write statistics, require not null.
-     */
-    private static void writeTotalRebalanceStatistics(
-        final Map<CacheGroupContext, Collection<RebalanceFuture>> rebFutrs,
-        final Map<ClusterNode, Integer> nodeAliases,
-        final StringJoiner joiner
-    ) {
-        assert nonNull(rebFutrs);
-        assert nonNull(nodeAliases);
-        assert nonNull(joiner);
+            long fe = supStat.fullEntries(), he = supStat.histEntries();
+            long fb = supStat.fullBytes(), hb = supStat.histBytes();
 
-        long minStartTime = minStartTime(toRebalanceFutureStream(rebFutrs));
-        long maxEndTime = maxEndTime(toRebalanceFutureStream(rebFutrs));
+            sb.a(supInfo(nodeId++, fp, hp, fe, he, fb, hb, supStat.start(), supStat.end()));
+        }
 
-        joiner.add("Total information (" + SUCCESSFUL_OR_NOT_REBALANCE_TEXT + "):")
-            .add("[" + toStartEndDuration(minStartTime, maxEndTime) + "]");
+        sb.a(SUP_STAT_ALIASES);
 
-        Map<ClusterNode, List<RebalanceMessageStatistics>> supplierStat =
-            toSupplierStatistics(toRebalanceFutureStream(rebFutrs));
+        nodeId = 0;
+        for (ClusterNode supNode : supStats.keySet())
+            sb.a(supInfo(nodeId++, supNode));
 
-        writeSupplierRebalanceStatistics(supplierStat, nodeAliases, joiner);
-    }
+        if (!availablePrintPartitionsDistribution(stat.end() - stat.start()))
+            return sb.toString();
```

Comment:
Why do we always accumulate partition statistics in SB if it may be not printed?

Reply:
I didn't understand the question.
In this lines we only check whether we need to print distribution partitions or not, if it is not necessary that output without it.

Reply:
Ok, got it